got-plt/README.md

```diff
@@ -0,0 +1,419 @@
+ld -shared -o libbasket.so basket.o
+ld -e_start -pie -dynamic-linker=/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 -rpath=. -o main start.o main.o -L. -lbasket
+```
+
+Running the program prints nothing, but that's not relevant or our investigation:
+```
+$ ./main
+```
+
+## Source Code Analysis
+
+The two source code files are simplistic.
+
+`basket.c` defines a global variable `basket_size` (initialized at `3`) and a function `flowers()`:
+```
+$ cat basket.c
+unsigned long basket_size = 3;
+
+void flowers(void)
+{
+	basket_size = 55;
+}
+```
+Both symbols (the global variable `basket_size` and the function `flowers()`) are exported (i.e. not marked `static`).
+
+`main.c` calls the `flowers()` function and writes to the `basket_size` variable:
+```
+$ cat main.c
+#include "basket.h"
+
+int main(void)
+{
+	flowers();
+	basket_size = 99;
+
+	return 0;
+}
+```
+`main.c` references the two symbols through their declarations in `basket.h`:
+```
+#ifndef BASKET_H_
+#define BASKET_H_	1
+
+extern unsigned long basket_size;
+
+void flowers(void);
+
+#endif
+```
+
+The `basket_size` and the `flowers` symbols will be part of the `libbasket.so` library.
+We expect the `main` program to access these symbols via GOT.
+We use static and dynamic analysis to validate our expectation.
+
+## Static Analysis
+
+**Note**: The addresses shown in the analysis will differ from your own.
+Different build tools, different environments, different versions of the standard C library, memory randomization will lead to other addresses being used.
+
+We use `objdump` and `readelf` for static binary analysis of the `main` executable file and the `libbasket.so` library file.
+
+We check the symbols in both files:
+```
+$ nm libbasket.so
+0000000000201018 D basket_size
+[...]
+0000000000000298 T flowers
+[...]
+$ nm main
+                 U basket_size
+[...]
+                 U flowers
+[...]
+0000000000000391 T main
+[...]
+```
+`libbasket.so` defines and exports the `basket_size` and the `flowers` symbols.
+`main` defines and exports the `main` symbol and uses the `basket_size` and `flowers`, both of which are undefined.
+Their addresses will only be known at runtime.
+
+We see that `main` has `.got` and `.got.plt` sections:
+```
+$ readelf -S main
+There are 18 section headers, starting at offset 0x13e0:
+
+Section Headers:
+  [Nr] Name              Type             Address           Offset
+       Size              EntSize          Flags  Link  Info  Align
+[...]
+  [12] .got              PROGBITS         0000000000200ff8  00000ff8
+       0000000000000008  0000000000000008  WA       0     0     8
+  [13] .got.plt          PROGBITS         0000000000201000  00001000
+       0000000000000020  0000000000000008  WA       0     0     8
+[...]
+```
+Both sections are writable and readable (`WA` in the `Flags` column, for *write* and *access*).
+`.got` starts at address `0x200ff8`, `.got.plt` starts at address `0x201000`.
+
```

Comment:
Maybe worth adding: "The values of these addresses will change at load time.".

Reply:
I mentioned at the beginning of the section.

**Note**: The addresses shown in the analysis will differ from your own.
Different build tools, different environments, different versions of the standard C library, memory randomization will lead to other addresses being used.

Do you think I should emphasize this?

I also added a mention in the dynamic analysis section.

Reply:
I was actually thinking about the difference between the address of `.got` and `.got.plt` in static vs. dynamic analysis (ex. `0x200ff8` vs. `0x555555754ff8`), not about the difference produced by local tool. But yes, the change it's pretty obvious.

Reply:
Oh, I get it. I will update that.

Reply:
I added a note in the dynamic analysis section:

Note that, because `main` is a position-independent executable, these addresses differ from the ones found when doing static analysis: `0x200ff8` and `0x201000`.
Even so, their relative placement and page address (the last 3 hex digits) are identical.